extensions/amp-ad-network-cloudflare-impl/0.1/amp-ad-network-cloudflare-impl.js

```diff
@@ -0,0 +1,109 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpA4A} from '../../amp-a4a/0.1/amp-a4a';
+import {base64UrlDecodeToBytes} from '../../../src/utils/base64';
+import {dev} from '../../../src/log';
+
+
+// Header that will contain Cloudflare generated signature
+/** @type {string} */
+const AMP_SIGNATURE_HEADER = 'X-AmpAdSignature';
+
+// Base URL of Ad server.  The extra /_a4a prefix signals Cloudflare
+// A4A header processing to generate a signature and CORS headers.
+/** @type {string} */
+//const CLOUDFLARE_BASE_URL = 'http://www.mycloudflaredomain.com/_a4a';
+
+// this points to the internal "fake" cloudflare signed Ad server.
+/** @type {string} */
+const CLOUDFLARE_BASE_URL =
+  '/extensions/amp-ad-network-cloudflare-impl/0.1/data';
+
+/**
+ * This is a minimalistic AmpA4A implementation that primarily gets an Ad
+ * through a source URL and extracts the Cloudflare generated signature
+ * from a HTTP header.  This is then given to A4A to validate against
+ * the cloudflare signing key.
+ */
+export class AmpAdNetworkCloudflareImpl extends AmpA4A {
+
+  /**
+   * Optional constructor for ad network implementation.
+   * Note: Can be removed if no logic is needed here.
+   *
+   * @param {!Element} element
+   */
+  /*
+  constructor(element) {
+    // Any initialization code here.
+    super(element);
+  }
+  */
+
+  /**
+   * Validate the tag parameters.  If invalid, ad ad will not be displayed.
+   * @override
+   */
+  isValidElement() {
+    return this.element.hasAttribute('src') && this.isAmpAdElement();
+  }
+
+  /** @override */
+  getSigningServiceNames() {
+    // this specifies verification for Cloudflare signing
+    return ['cloudflare'];
+  }
+
+  /** @override */
+  getAdUrl() {
+    // You can optionally use data from this context to augment the URL,
+    // to pass data back to the network, to aid in ad selection.  For example:
+    //   const rect = this.getIntersectionElementLayoutBox();
+    // can extract the layout size and be used to pass layout hints to the server.
+
+    // Be sure that this doesn't allow the generation of ANY url, since it could
+    // allow cookie leaks, etc.
+
+    return CLOUDFLARE_BASE_URL + this.element.getAttribute('src');
+  }
+
+  /**
+   * Extract creative and signature from a Cloudflare signed response.
+   *
+   * Note: Invalid A4A content will NOT have a signature, which will automatically
+   *   cause the A4A processing to fall through to the 3p ad flow.
```

Comment:
Note that lack of signature does not actually cause the 3p ad flow to be used where a xdomain iframe is created in which custom JS executes to generate an ad request.  Instead, it will take the creative given and render it directly within a cross domain frame using either network cache (if Android) or safeframe (if iOS).